### PR TITLE
ft-166240783 allow user to view his profile

### PR DESCRIPTION
**Pull request by alainburindi**
#### What does this PR do?
users profile

#### How should this be manually tested?
* Login to Authors haven and you will be redirected to your profile

### screenshots
Screen Shot 2019-09-04 at 12 51 44

Screen Shot 2019-09-04 at 12 52 46

Screen Shot 2019-09-04 at 12 52 51

Screen Shot 2019-09-04 at 12 52 55

Screen Shot 2019-09-04 at 12 53 05

Screen Shot 2019-09-04 at 12 53 26

### CODE REVIEW

File: src/App/Components/Login/index.js [hunk 1]
```diff
@@ -35,9 +35,12 @@ export class Login extends Component {
...
 
     let errorMessage;
-    const { authReducer: { errors, isAuthorized } = {}, history } = this.props;
+    const {
+      authReducer: { errors, isAuthorized, user },
+      history
+    } = this.props;
     if (!isAuthorized && errors) errorMessage = errors.error || errors.email;
-    if (isAuthorized) history.replace('/');
+    if (isAuthorized) history.replace(`/profile/${user.username}`);
```
Comment (nkpremices) [inline src/App/Components/Login/index.js:43]: @alainburindi is this not reloading the store?

File: src/App/Components/Profile/BookMark.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+import _ from 'lodash';
```
Comment (nkpremices) [inline src/App/Components/Profile/BookMark.js]: @alainburindi `lodash` is a very huge module. please, import only what you need from it.
Like this `import isEmpty from 'lodash/isEmpty';`

File: src/App/Components/Profile/BookMark.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import ReactImageFallback from 'react-image-fallback';
+import propTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+
+export default function BookMark(props) {
+  const { data, remove, owner } = props;
+
+  return !_.isEmpty(data) ? (
+    <div>
+      {data.bookmarks.map(bookmark => (
```
Comment (nkpremices) [inline src/App/Components/Profile/BookMark.js:17]: @alainburindi if you don't mind, you could improve here by destructuring `data`

File: src/App/Components/Profile/BookMark.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        <div key={bookmark.Article.id} className="row mt-4 bookmark_container p-3">
```
Comment (nkpremices) [inline src/App/Components/Profile/BookMark.js:18]: Same here @alainburindi

Comment (nkpremices) [reply src/App/Components/Profile/BookMark.js:18]: There are many iterations of this actually.

File: src/App/Components/Profile/Follower.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import _ from 'lodash';
```
Comment (nkpremices) [inline src/App/Components/Profile/Follower.js]: Once again, import only what you need @alainburindi

File: src/App/Components/Profile/BookMark.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+              <div className={`category ${bookmark.Article.Category.name}`} />
+              <p className="col-3">{bookmark.Article.Category.name}</p>
+            </div>
+          </div>
+          <p className="col-12 m-3 bookmark_container_description">
+            {bookmark.Article.description}
+          </p>
+          <ReactImageFallback
+            src={bookmark.Article.User.profileImage}
+            fallbackImage="https://res.cloudinary.com/dkqqm2qwn/image/upload/v1566378272/blank-profile-picture-973460_960_720_jyfolk.png"
```
Comment (nkpremices) [inline src/App/Components/Profile/BookMark.js]: we already have a file in `urls/configs` for all urls @alainburindi

File: src/App/Components/Profile/Follower.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    following,
+    owner
+  } = props;
+  return !_.isEmpty(followers) ? (
+    <div>
+      {followers.map(follower => (
+        <div key={follower.follower} className="row follower_container p-2 mb-2">
+          <ReactImageFallback
+            src={follower.profileImage}
+            fallbackImage="https://res.cloudinary.com/dkqqm2qwn/image/upload/v1566378272/blank-profile-picture-973460_960_720_jyfolk.png"
```
Comment (nkpremices) [inline src/App/Components/Profile/Follower.js]: In the `urls` file @alainburindi

File: src/App/Components/Profile/Following.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    owner
+  } = props;
+
+  return !_.isEmpty(following) ? (
+    <div key="kkkk">
+      {following.map(follow => (
+        <div key={follow.following} className="row follower_container p-2 mb-2">
+          <ReactImageFallback
+            src={follow.profileImage}
+            fallbackImage="https://res.cloudinary.com/dkqqm2qwn/image/upload/v1566378272/blank-profile-picture-973460_960_720_jyfolk.png"
```
Comment (nkpremices) [inline src/App/Components/Profile/Following.js]: In the `urls` file @alainburindi

File: src/App/Components/Profile/Following.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+export default function Following(props) {
+  const {
+    following: { following },
+    unfollow,
+    owner
+  } = props;
+
+  return !_.isEmpty(following) ? (
+    <div key="kkkk">
```
Comment (nkpremices) [inline src/App/Components/Profile/Following.js]: @alainburindi This key ?

File: src/App/Components/Profile/Profile.js [hunk 10]
```diff
@@ -0,0 +1,392 @@
+import React, { Component } from 'react';
+import propTypes from 'prop-types';
+import ReactImageFallback from 'react-image-fallback';
+import SimpleReactValidator from 'simple-react-validator';
+import { connect } from 'react-redux';
+import numeral from 'numeral';
+import _ from 'underscore';
```
Comment (nkpremices) [inline src/App/Components/Profile/Profile.js]: Underscore is a deprecated version of lodash @alainburindi .
Import lodash instead please

File: src/App/Components/Profile/Profile.js [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+      follower,
+      following,
+      bookmarks,
+      owner,
+      follow: followAction,
+      deleteArticle: deleteArticleAction
+    } = this.props;
+
+    const { temporally } = this.state;
+    return !_.isEmpty(profile) && follower && bookmarks && following ? (
```
Comment (nkpremices) [inline src/App/Components/Profile/Profile.js]: Again, import what you need @alainburindi

File: src/App/Components/Profile/Profile.js [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+      <div>
```
Comment (nkpremices) [inline src/App/Components/Profile/Profile.js:178]: I believe that destructuring would help improve this component @alainburindi

File: src/App/Components/Profile/Profile.js [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+        <NavBar />
+        <div className="container">
+          <div className="row">
+            <div className="col-md-3">
+              <div className="profile-container row p-0" style={{ marginTop: '48px' }}>
+                <div className="profile-container__cover">
+                  <ReactImageFallback
+                    src={profile.profileImage}
+                    fallbackImage="https://res.cloudinary.com/dkqqm2qwn/image/upload/v1566378272/blank-profile-picture-973460_960_720_jyfolk.png"
```
Comment (nkpremices) [inline src/App/Components/Profile/Profile.js]: in the urls file @alainburindi

File: src/App/Components/Profile/SingleArticle.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+import Delete from './DeleteModal';
+import countRating from '../../../Helpers/countRating';
+
+export default function SingleArticle(props) {
+  const { article, owner, deleteArticleAction } = props;
+  return (
+    <div key={article.id} className="row single-article mt-4">
+      <div className="single-article-header mt-3 row col-12">
+        <p className="col-11 single-article-header-left">&nbsp;&nbsp;&nbsp;&nbsp;</p>
+        {/* delete modal */}
```
Comment (nkpremices) [inline src/App/Components/Profile/SingleArticle.js:17]: commented code @alainburindi

**Review by nkpremices — CHANGES_REQUESTED** [review]

**Comment by nkpremices** [comment]
great job @alainburindi 
I left some comments on your PR

File: src/App/Components/Login/index.js (hunk 1 above)
Comment (alainburindi) [reply src/App/Components/Login/index.js:43]: nope

File: src/App/Components/Profile/BookMark.js (hunk 3 above)
Comment (alainburindi) [reply src/App/Components/Profile/BookMark.js:17]: I did it at line 8

File: src/App/Components/Profile/BookMark.js (hunk 4 above)
Comment (alainburindi) [reply src/App/Components/Profile/BookMark.js:18]: I can not bookmark as you can see at line 33 I still need the bookmark object

File: src/App/Components/Profile/BookMark.js (hunk 6 above)
Comment (alainburindi) [reply src/App/Components/Profile/BookMark.js]: didn't know

File: src/App/Components/Profile/SingleArticle.js (hunk 14 above)
Comment (alainburindi) [reply src/App/Components/Profile/SingleArticle.js:17]: that is a comment

File: src/App/Components/Profile/Profile.js (hunk 12 above)
Comment (alainburindi) [reply src/App/Components/Profile/Profile.js:178]: let me see what I can improve

**Review by nkpremices — APPROVED** [review]
Great @alainburindi

File: src/App/ApiActions/getRating.js [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import axios from '../../configs/axios';
+
+export default async slug => axios
```
Comment (MemunaHaruna) [inline src/App/ApiActions/getRating.js:3]: @alainburindi what's going on here? Why are you not using the error messages gotten from the server?

File: src/App/ApiActions/getStatistics.js [hunk 16]
```diff
@@ -0,0 +1,6 @@
+import axios from '../../configs/axios';
+
+export default async slug => axios
+  .get(`/api/articles/${slug}/stats`)
```
Comment (MemunaHaruna) [inline src/App/ApiActions/getStatistics.js:4]: @alainburindi same here

File: src/App/Components/Profile/BookMark.js (hunk 4 above)
Comment (MemunaHaruna) [reply src/App/Components/Profile/BookMark.js:18]: @nkpremices can you sync with @alainburindi and show him an example of what you mean?

File: src/App/ApiActions/getRating.js (hunk 15 above)
Comment (alainburindi) [reply src/App/ApiActions/getRating.js:3]: The error message is coming from the backend when the article doesn't have any available statistics which is normal. then I was returning `0` to make sure I'm handling all possible cases. then I was checking if the value is equal to 0 or not. Do you suggest anything else??

File: src/App/ApiActions/getStatistics.js (hunk 16 above)
Comment (alainburindi) [reply src/App/ApiActions/getStatistics.js:4]: the same here

File: src/App/Components/Profile/BookMark.js (hunk 4 above)
Comment (alainburindi) [reply src/App/Components/Profile/BookMark.js:18]: I explained to @nkpremices the reason and I have destructured where it was possible

File: src/App/ApiActions/getRating.js (hunk 15 above)
Comment (MemunaHaruna) [reply src/App/ApiActions/getRating.js:3]: @alainburindi my actual question is: why do you need this file at all? Why not make the api call as a Redux action and dispatch appropriate messages to the reducer based on responses from the server?

**Review by MemunaHaruna — COMMENTED** [review]
@alainburindi Good work. I left a new comment for you. Two things:
- Rebase and resolve conflicts
- From the bookmarks screenshots, the profile pictures aren't the same shape. Compare the picture of Premices and the other avatars.

Finally, did you implement pagination on the Bookmark page?

File: src/App/ApiActions/getRating.js (hunk 15 above)
Comment (alainburindi) [reply src/App/ApiActions/getRating.js:3]: While fetching the user's profile I was getting his articles but each article must have his own statistics. So I had to map from the response and get appropriate statistics, push them to the corresponding article and dispatch after

**Review by MemunaHaruna — APPROVED** [review]

**Pull request merged**